stdlib/public/runtime/ImageInspectionInit.cpp

```diff
@@ -0,0 +1,33 @@
+//===-- ImageInspectionInit.cpp ---------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file along with swift_sections.S is prepended to each shared library
+// on an ELF target which contains protocol and metadata sections.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__) || defined(__ANDROID__)
+
+#include "ImageInspection.h"
+
+using namespace swift;
+
+// This is called at startup and by each shared object as it is dlopen()'d to
+// allow the section data for the object to be loaded
+__attribute__((constructor))
+static void
+loadSectionData() {
+  addImageProtocolConformanceBlock(&__swift2_protocol_conformances_block);
+  addImageTypeMetadataRecordBlock(&__swift2_type_metadata_block);
+}
```

Comment:
This is doing too much work, since it will force initialization of the lookup caches even if they haven't been enabled yet, and is not robust ABI if `libswiftCore.so` changes and requires new per-image registration hooks in the future. It would be better for the runtime to export an entry point `swift_registerLoadedImage(const void *image)`, and have the static constructor do nothing but call that with a handle to the loaded image, something like:

```
SWIFT_RUNTIME_EXPORT extern "C" void swift_registerLoadedImage(const void *dsoHandle);

__attribute__((constructor))
static void onLoad() {
  swift_registerLoadedImage(__dso_handle);
}
```

The runtime implementation of `swift_registerLoadedImage` should be responsible for extracting and registering the sections that parts of the runtime are interested in, and do so only if those parts have already been initialized. It should be as close to a no-op as possible if an image is loaded before any lookup tables are initialized.

We should also make sure that this static constructor is not linked into executables or into `libswiftCore.so` itself, since the executable and runtime by their nature have to have been loaded before any lookup caches are triggered.